notify/impl.go

```diff
@@ -842,75 +839,86 @@ func (n *Wechat) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
 	if !ok {
 		return false, fmt.Errorf("group key missing")
 	}
-	data := n.tmpl.Data(receiverName(ctx, n.logger), groupLabels(ctx, n.logger), as...)
+
 	level.Debug(n.logger).Log("msg", "Notifying Wechat", "incident", key)
+	data := n.tmpl.Data(receiverName(ctx, n.logger), groupLabels(ctx, n.logger), as...)
 
 	var err error
 	tmpl := tmplText(n.tmpl, data, &err)
-
-	var (
-		msg    interface{}
-		apiURL string
-		apiMsg = weChatMessage{
-			Content: tmpl(n.conf.Message),
-		}
-		alerts = types.Alerts(as...)
-	)
-	parameters := url.Values{}
-	parameters.Add("corpsecret", tmpl(string(n.conf.APISecret)))
-	parameters.Add("corpid", tmpl(string(n.conf.CorpID)))
-	apiURL = n.conf.APIURL + "gettoken"
-	u, err := url.Parse(apiURL)
-	if err != nil {
-		return false, err
-	}
-	u.RawQuery = parameters.Encode()
-	level.Debug(n.logger).Log("msg", "Sending Wechat  message", "incident", key, "url", u.String())
-	resp, err := ctxhttp.Get(ctx, http.DefaultClient, u.String())
 	if err != nil {
-		return true, err
-	}
-	defer resp.Body.Close()
-	var wechatToken WechatToken
-	if err := json.NewDecoder(resp.Body).Decode(&wechatToken); err != nil {
 		return false, err
 	}
-	postMessageURL := n.conf.APIURL + "message/send?access_token=" + wechatToken.AccessToken
-	switch alerts.Status() {
-	case model.AlertResolved:
-		msg = &weChatCloseMessage{Text: apiMsg,
-			ToUser:  tmpl(n.conf.ToUser),
-			ToParty: tmpl(n.conf.ToParty),
-			Totag:   tmpl(n.conf.ToTag),
-			AgentID: tmpl(n.conf.AgentID),
-			Type:    "text",
-			Safe:    "0"}
-	default:
-		msg = &weChatCreateMessage{
-			Text: weChatMessage{
-				Content: tmpl(n.conf.Message),
-			},
-			ToUser:  tmpl(n.conf.ToUser),
-			ToParty: tmpl(n.conf.ToParty),
-			Totag:   tmpl(n.conf.ToTag),
-			AgentID: tmpl(n.conf.AgentID),
-			Type:    "text",
-			Safe:    "0",
+
+	var accessToken string
+	// Refresh AccessToken over 2 hours
+	if n.accessToken != "" && time.Now().Sub(n.accessTokenAt) < 110*time.Minute {
```

Comment:
You can change `110*time.Minute` to `2*time.Hour`

Reply:
This could be changed to

```go
if n.accessToken == "" || time.Now().Sub(n.accessTokenAt) > 2*time.Hour {
        // refresh token
        // Set n.accessToken and n.AccessTokenAt
}

// Down below, use n.accessToken directly
```